pkg/provider/loadbalancer/securitygroup/securitygroup.go

```diff
@@ -0,0 +1,448 @@
+	sg       *network.SecurityGroup
+	snapshot []byte
+
+	// helper map to store the security rules
+	// name -> security rule
+	rules      map[string]*network.SecurityRule
+	priorities map[int32]string
+}
+
+func NewSecurityGroupHelper(sg *network.SecurityGroup) (*RuleHelper, error) {
+	if sg == nil ||
+		sg.Name == nil ||
+		sg.SecurityGroupPropertiesFormat == nil ||
+		sg.SecurityRules == nil {
+		// In the real world, this should never happen.
+		// Anyway, defensively check it.
+		return nil, ErrInvalidSecurityGroup
+	}
+	var (
+		logger     = klog.Background().WithName("RuleHelper").WithValues("security-group-name", ptr.To(sg.Name))
+		rules      = make(map[string]*network.SecurityRule, len(*sg.SecurityGroupPropertiesFormat.SecurityRules))
+		priorities = make(map[int32]string, len(*sg.SecurityGroupPropertiesFormat.SecurityRules))
+	)
+	for i := range *sg.SecurityGroupPropertiesFormat.SecurityRules {
+		rule := (*sg.SecurityGroupPropertiesFormat.SecurityRules)[i]
+		rules[*rule.Name] = &rule
+		priorities[*rule.Priority] = *rule.Name
+	}
+
+	snapshot := makeSecurityGroupSnapshot(sg)
+
+	return &RuleHelper{
+		logger: logger,
+		sg:     sg,
+
+		rules:      rules,
+		priorities: priorities,
+		snapshot:   snapshot,
+	}, nil
+}
+
+type rulePriorityPrefer string
+
+const (
+	rulePriorityPreferFromStart rulePriorityPrefer = "from_start"
+	rulePriorityPreferFromEnd   rulePriorityPrefer = "from_end"
+)
+
+func (helper *RuleHelper) nextRulePriority(prefer rulePriorityPrefer) (int32, error) {
+	var (
+		init, end = consts.LoadBalancerMinimumPriority, consts.LoadBalancerMaximumPriority
+		delta     = 1
+	)
+	if prefer == rulePriorityPreferFromEnd {
+		init, end, delta = end-1, init-1, -1
+	}
+
+	for init != end {
+		p := int32(init)
+		if _, found := helper.priorities[p]; found {
+			init += delta
+			continue
+		}
+		return p, nil
+	}
+
+	return 0, ErrSecurityRulePriorityExhausted
+}
+
+func (helper *RuleHelper) getOrCreateRule(name string, priorityPrefer rulePriorityPrefer) (*network.SecurityRule, error) {
+	logger := helper.logger.WithName("getOrCreateRule").WithValues("rule-name", name)
+
+	if rule, found := helper.rules[name]; found {
+		logger.V(4).Info("Found an existing rule", "priority", *rule.Priority)
+		return rule, nil
+	}
+
+	priority, err := helper.nextRulePriority(priorityPrefer)
+	if err != nil {
+		// NOTE: right now it won't happen because the number of rules is limited.
+		//       maxPriority[4096] - minPriority[500] > maxSecurityRulesPerGroup[1000]
+		helper.logger.Error(err, "Failed to get an available rule priority")
+		return nil, err
+	}
+	rule := &network.SecurityRule{
+		Name: ptr.To(name),
+		SecurityRulePropertiesFormat: &network.SecurityRulePropertiesFormat{
+			Priority: ptr.To(priority),
+		},
+	}
+
+	helper.rules[name] = rule
+	helper.priorities[priority] = name
+
+	logger.V(4).Info("Adding a new rule", "rule-name", name, "priority", priority)
+
+	return rule, nil
+}
+
+func (helper *RuleHelper) addAllowRule(
```

Comment:
Could you add comments for newly introduced functions and packages?

Reply:
Added.